.github/workflows/php-linting.yml

```diff
@@ -0,0 +1,217 @@
+
+jobs:
+  changed_files:
+    name: detect changed files
+    runs-on: ubuntu-latest
+    outputs:
+      php: ${{ steps.filter.outputs.php }}
+      misc: ${{ steps.filter.outputs.misc }}
+      php_excluded_files: ${{ steps.filter2.outputs.php_excluded_files }}
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          list-files: json
+          filters: |
+            php:
+              # If any PHP file changed, they need checking.
+              - added|modified:
+                - '**.php'
+            misc:
+              # If composer or phpcs config changed, there may be a new standard.
+              - 'composer.json'
+              - 'composer.lock'
+              - '.phpcs.config.xml'
+              - '.phpcs.xml.dist'
+              # If the excludelist changed, run to ensure newly non-excluded files pass.
+              - 'bin/phpcs-excludelist.txt'
+              # If other files used by this workflow changed, run it to test those changes.
+              - 'bin/parallel-lint.sh'
+              - '.github/files/php-linting-phpcs.xml'
+              - '.github/matchers/phpcs-problem-matcher.json'
+              - '.github/matchers/php-lint-problem-matcher.json'
+              - '.github/workflows/php-linting.yml'
+
+      - id: filter2
+        shell: bash
+        run: |
+          printf '::set-output name=php_excluded_files::%s\n' $(
+          jq -r '.[]' <<'EOF' | grep -Fx -f bin/phpcs-excludelist.txt | jq -nRc '[ inputs ]'
+          ${{ steps.filter.outputs.php_files }}
+          EOF
+          )
+
+  lint:
+    name: PHP lint (${{ matrix.php-versions }})
+    runs-on: ubuntu-latest
+    needs: changed_files
+    if: needs.changed_files.outputs.php == 'true' || needs.changed_files.outputs.misc == 'true'
+    continue-on-error: ${{ matrix.experimental }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        php-versions: [ '5.6', '7.0', '7.4' ]
+        experimental: [ false ]
+
+        include:
+          - php-versions: '8.0'
+            experimental: true
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-versions }}
+          extensions: intl, json, mbstring, xml
+
+      - name: Get Composer cache directory
+        id: composer-cache
+        run: |
+          echo "::set-output name=dir::$(composer config cache-files-dir)"
+      - uses: actions/cache@v2
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-composer-
+
+      - name: Tool versions
+        run: |
+          php --version
+          composer --version
+
+      - name: Install dependencies
+        run: |
+          # Install stuff ignoring platform reqs.
+          composer install --ignore-platform-reqs
```

Comment:
Should we run the remove before the install?

Reply:
The remove does an implicit update, which can wind up installing wrong versions of packages (cf. #17259). IIRC a `composer install` also avoids the problem by more closely sticking to the lock file.

We could do the remove first with the `COMPOSER_ROOT_VERSION=dev-master` that was done in #17259 if you really want.